TLC5957.cpp

```diff
@@ -0,0 +1,366 @@
+#include "TLC5957.h"
+#include <SPI.h>
+#include <Arduino.h>
+
+#define GS_BITS 16
+#define BC_BITS 3
+#define CC_BITS 9
+#define FC_BITS 48
+
+#define WRTFC 5
+#define FCWRTEN 15
+#define WRTGS 1
+#define LATGS 3
+
+void TLC5957::init(uint8_t lat, uint8_t sin, uint8_t sclk, uint8_t gsclk)
+{
+    _lat = lat;
+    _sin = sin;
+    _sclk = sclk;
+    _gsclk = gsclk;
+
+    pinMode(_lat, OUTPUT);
+    pinMode(_sin, OUTPUT);
+    pinMode(_sclk, OUTPUT);
+    pinMode(_gsclk, OUTPUT);
+
+    mSettings = SPISettings(_spi_baud_rate, MSBFIRST, SPI_MODE0);
+    SPI.setMOSI(_sin);
+    SPI.begin();
+
+    digitalWrite(_lat, LOW);
+    setGsclkFreq(_gsclk_frequency);
+}
+
+void TLC5957::setSpiBaudRate(uint32_t baud_rate)
+{
+    _spi_baud_rate = baud_rate;
+    mSettings = SPISettings(_spi_baud_rate, MSBFIRST, SPI_MODE0);
+}
+
+uint32_t TLC5957:: getSpiBaudRate()
+{
+    return _spi_baud_rate;
+}
+
+void TLC5957::setGsclkFreq(uint32_t gsclk_frequency)
+{
+    _gsclk_frequency = gsclk_frequency;
+    analogWriteFrequency(_gsclk, _gsclk_frequency);
+    analogWriteResolution(1);
+    analogWrite(_gsclk, 1);
+}
+
+uint32_t TLC5957::getGsclkFreq()
+{
+    return _gsclk_frequency;
+}
+
+void TLC5957::latch(int num_edges)
+{
+    // TODO: do I  need to enforce integer division??
+    int lat_delay_microseconds = _spi_baud_rate / 1'000'000;
+    delayMicroseconds(lat_delay_microseconds);
+    digitalWrite(_lat, HIGH);
+    delayMicroseconds(lat_delay_microseconds * num_edges);
+    digitalWrite(_lat, LOW);
+    delayMicroseconds(lat_delay_microseconds);
+}
+
+void TLC5957::latch(uint16_t data, uint8_t data_len, uint8_t num_edges)
+{
+    SPI.end();
+    digitalWrite(_sclk, LOW);
+    for (uint8_t i = data_len - 1; i >= 0; i--)
+    {
+        digitalWrite(_sin, data >> i & 1);
+        digitalWrite(_sclk, HIGH);
+        digitalWrite(_sclk, LOW);
+        if (i == num_edges - 1)
```

Comment:
you should add comments to where in the datasheet you are pulling this information from. This seemed to require some design.